client/network/src/protocol/generic_proto/behaviour.rs

```diff
@@ -873,212 +1132,277 @@ impl NetworkBehaviour for GenericProto {
+						opening_and_closing, closed, closing, banned_until
+					};
 				}
-			}
-			_ => {}
-		}
-	}
+			},
 
-	fn inject_disconnected(&mut self, peer_id: &PeerId) {
-		match self.peers.remove(peer_id) {
-			None | Some(PeerState::Requested) | Some(PeerState::PendingRequest { .. }) |
-			Some(PeerState::Banned { .. }) =>
-				// This is a serious bug either in this state machine or in libp2p.
-				error!(target: "sub-libp2p",
-					"`inject_disconnected` called for unknown peer {}",
-					peer_id),
+			// DisabledPendingEnable => DisabledPendingEnable | Banned
+			PeerState::DisabledPendingEnable { mut opening_and_closing, mut closed, mut closing, timer_deadline, timer } => {
+				debug!(
+					target: "sub-libp2p",
+					"Libp2p => Disconnected({}, {:?}): Disabled but pending enable.",
+					peer_id, *conn
+				);
 
-			Some(PeerState::Disabled { open, banned_until, .. }) => {
-				if !open.is_empty() {
+				if let Some(pos) = opening_and_closing.iter().position(|c| *c == *conn) {
+					opening_and_closing.remove(pos);
+				} else if let Some(pos) = closed.iter().position(|c| *c == *conn) {
+					closed.remove(pos);
+				} else if let Some(pos) = closing.iter().position(|c| *c == *conn) {
+					closing.remove(pos);
+				} else {
 					debug_assert!(false);
-					error!(
-						target: "sub-libp2p",
-						"State mismatch: disconnected from {} with non-empty list of connections",
-						peer_id
-					);
+					error!(target: "sub-libp2p",
+						"inject_connection_closed: State mismatch in the custom protos handler");
 				}
-				debug!(target: "sub-libp2p", "Libp2p => Disconnected({}): Was disabled.", peer_id);
-				if let Some(until) = banned_until {
-					self.peers.insert(peer_id.clone(), PeerState::Banned { until });
+
+				if opening_and_closing.is_empty() && closed.is_empty() && closing.is_empty() {
+					debug!(target: "sub-libp2p", "PSM <= Dropped({})", peer_id);
+					self.peerset.dropped(peer_id.clone());
+					*entry.get_mut() = PeerState::Banned { until: timer_deadline };
+
+				} else {
+					*entry.get_mut() = PeerState::DisabledPendingEnable {
+						opening_and_closing, closed, closing, timer_deadline, timer
+					};
 				}
-			}
+			},
 
-			Some(PeerState::DisabledPendingEnable { open, timer_deadline, .. }) => {
-				if !open.is_empty() {
+			// Incoming => Incoming | Disabled | Banned | Ø
+			PeerState::Incoming { mut opening_and_closing, mut closing, mut closed, mut open_desired, banned_until } => {
+				debug!(
+					target: "sub-libp2p",
+					"Libp2p => Disconnected({}, {:?}): OpenDesired.",
+					peer_id, *conn
+				);
+
+				if let Some(pos) = opening_and_closing.iter().position(|c| *c == *conn) {
+					opening_and_closing.remove(pos);
+				} else if let Some(pos) = closing.iter().position(|c| *c == *conn) {
+					closing.remove(pos);
+				} else if let Some(pos) = closed.iter().position(|c| *c == *conn) {
+					closed.remove(pos);
+				} else if let Some(pos) = open_desired.iter().position(|c| *c == *conn) {
```

Comment:
I went for the approach of multiple fields because the list of fields isn't the same for all the variants of `PeerState`.
In particular, the `open_desired` field is only present in the `Incoming` state, and the `open` and `opening` fields are only present in `Enabled`.

Reply:
I understand, though I think it is a trade-off for the worse in terms of code complexity and possibilities for mistakes. Asserting that some connection states are never used by some peer states seems easier to me than making sure that this pop/push scheme between all these different vectors for state transitions is sound. I may be wrong, of course, as one usually only knows for sure after having actually tried it. Not to mention there is also the option of multiple state enums to precisely model the possible connection states in the different peer states.

Reply:
I'll do the change, but after I've figured out why the tests are failing.